### PR TITLE
✨ Defaults for external code parameterframes

### PR DESCRIPTION
**Pull request by je-cook**
## Description

Add defaults for codes. Needs some finessing 

## Interface Changes

If you've had to update an interface or introduce a new interface as part of your change then let us know here.

## Checklist

I confirm that I have completed the following checks:

- [ ] Tests run locally and pass `pytest tests --reactor`
- [ ] Code quality checks run locally and pass `flake8` and `black .`
- [ ] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

File: bluemira/codes/params.py [hunk 1]
```diff
@@ -159,3 +160,25 @@ def __setattr__(self, attr: str, value: Union[bool, str]):
...
+    """
+    new_param_dict = {}
+    for map_name, param_map in mappings.items():
+        if param_map.send:
+            new_param_dict[map_name] = {
+                "value": getattr(defaults, param_map.name),
+                "unit": param_map.unit,
+            }
+        else:
+            new_param_dict[map_name] = {"value": 0, "unit": param_map.unit}
```
Comment (je-cook) [inline bluemira/codes/params.py:182]: I dont really like this default but I couldn't think of a better option atm. Possibly at least have the same type

File: bluemira/codes/process/_solver.py [hunk 2]
```diff
@@ -141,7 +142,18 @@ def __init__(self, params: Union[Dict, ParameterFrame], build_config: BuildConfi
...
+        # Get process defaults from template file
+        mm = _make_writer(self.template_in_dat).data
+        process_inputs = make_dataclass("ProcesInputs", list(mm.keys()))(
+            {key: value.get_value for key, value in mm.items()}
+        )
+
+        self.params = make_mapped_default_parameter_frame(
+            process_inputs, mappings, ProcessSolverParams
+        )
+        self.params.update_from_frame(params)
```
Comment (je-cook) [inline bluemira/codes/process/_solver.py:154]: this is also not great as we end up reading in the template twice. Here and in setup.py. Wanted to keep it separate from setup though...

File: bluemira/codes/interface.py [hunk 3]
```diff
@@ -187,7 +187,7 @@ def _map_external_outputs_to_bluemira_params(
             if output_value is None or mapping.unit is None:
                 continue
             value = raw_uc(
-                output_value, getattr(self.params, bm_name).unit, mapping.unit
+                output_value, mapping.unit, getattr(self.params, bm_name).unit
```
Comment (je-cook) [inline bluemira/codes/interface.py:190]: Oops :)

**Review by je-cook — COMMENTED** [review]
I went though and changed the default units in a few places. The tests weren't failing in most cases before or after but at least I think we're nearly consistent now.

**Comment by je-cook** [comment]
The warning fix is in #1464 . I wasnt planning on fixing it, instead just waiting on that

File: bluemira/codes/process/_inputs.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    icc: List[int] = field(default_factory=lambda: [1, 2, 5, 8, 11, 13, 15, 16, 24, 25,
+                                                    26, 27, 30, 31, 32, 33, 34, 35, 36,
+                                                    60, 62, 65, 68, 72])
+    ixc: List[int] = field(default_factory=lambda: [2, 3, 4, 5, 6, 9, 13, 14, 16, 18,
+                                                    29, 36, 37, 38, 39, 41, 42, 44, 48,
+                                                    49, 50, 51, 52, 53, 54, 56, 57, 58,
+                                                    59, 60, 61, 102, 103, 106, 109, 110,
+                                                    113, 117, 122, 123])
+    # fmt: on
+    abktflnc: float = 15.0
```
Comment (hsaunders1904) [inline bluemira/codes/process/_inputs.py:70]: Ahh PROCESS, this is a real research science variable name 😆. Can't even begin to think what this means.

Comment (hsaunders1904) [reply bluemira/codes/process/_inputs.py:70]: Bit of a pain, I know, but should these have units 😨? Or we, at least, need a something in the docstring(s) that says where to find the units

File: bluemira/codes/process/_inputs.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    vftf: float = 0.3
+    vgap2: float = 0.05
+    vvblgap: float = 0.02
+    walalw: float = 8.0
+    zeffdiv: float = 3.5
+    zref: List[float] = field(
+        default_factory=lambda: [3.6, 1.2, 1.0, 2.8, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
+    )
+
+    def __iter__(self) -> Generator[field, None, None]:
```
Comment (hsaunders1904) [inline bluemira/codes/process/_inputs.py:261]: `field` is a function, I think. We may want `Field`, but aren't they `INVariable`s?

**Review by hsaunders1904 — CHANGES_REQUESTED** [review]
Still working my way through, but have some comments to start

File: bluemira/codes/process/api.py [hunk 6]
```diff
@@ -80,9 +79,42 @@ def __init__(self, filename):
     PROCESS_DICT = get_dicts()
 
 
-DEFAULT_INDAT = os.path.join(
-    get_bluemira_path("codes/process"), "PROCESS_DEFAULT_IN.DAT"
-)
+@dataclass
+class INVariable:
```
Comment (hsaunders1904) [inline bluemira/codes/process/api.py:83]: Could/should this be a private class?

```suggestion
class _INVariable:
```

File: bluemira/codes/process/api.py [hunk 7]
```diff
@@ -80,9 +79,42 @@ def __init__(self, filename):
...
+    """
+    Process io.in_dat.INVariable replica
+
+    Used to simulate what process does to input variables
+    for the InDat input file writer
+    """
```
Comment (hsaunders1904) [inline bluemira/codes/process/api.py:89]: Might be worth noting _why_ we need to simulate process in-variables, I remember from our conversations that this wasn't a straight-forward decision. Putting down some of that reasoning might stop has re-hashing the same conversations

File: bluemira/codes/process/params.py [hunk 8]
```diff
@@ -325,11 +326,40 @@ class ProcessSolverParams(MappedParameterFrame):
...
-        if self._mappings is None:
-            self._mappings = deepcopy(mappings)
         return self._mappings
+
+    @property
+    def defaults(self) -> ProcessInputs:
+        """
+        Default values for Process
+        """
+        return self._defaults
```
Comment (hsaunders1904) [inline bluemira/codes/process/params.py:342]: If these are edited, are they edited on the class? Do we need a copy, or, like I mentioned in a comment on the base class, could return a dictionary?

File: bluemira/codes/process/params.py [hunk 9]
```diff
@@ -325,11 +326,40 @@ class ProcessSolverParams(MappedParameterFrame):
...
+            "enbeam", "etanbi", "emult", "shldlth", "pnetelmw", "kappa95", "triang95",
+            "plascur/1d6", "powfmw", "pdt", "pdd", "pdivt", "pcoreradmw", "pedgeradmw",
+            "pedgeradmw", "pradmw", "plinepv*vol", "psyncpv*vol", "pbrempv*vol",
+            "bootipf", "betap", "taueff", "vburn", "fwith", "fwoth", "dr_tf_wp", "wwp1",
+            "tfinsgap", "r_cp_top", "rtfin", "r_tf_inboard_mid", "r_ts_ib_in",
+            "r_vv_ib_in", "r_fw_ib_in", "r_fw_ob_in", "r_vv_ob_in", "r_tf_outboard_mid",
+            "tfbusres", "ztot", "estotftgj", "tflegres", "pinjmw", "qss/1.0D6",
+            "thshield", "bmaxtfrp", "q95", "zeff", "vol", "rli", "faccd", "tfthko",
+            "h_cp_top", "hmax", "r_tf_inboard_out",
+        ]
```
Comment (hsaunders1904) [inline bluemira/codes/process/params.py:361]: Could be a module-level constant

File: examples/radiation_transport/heat_flux_calculation_DN.ipynb [hunk 10]
```diff
@@ -51,13 +51,8 @@
...
         "read_path = get_bluemira_path(\"radiation_transport/test_data\", subfolder=\"tests\")\n",
         "fw_name = \"DN_fw_shape.json\"\n",
-        "<<<<<<< HEAD\n",
-        "fw_name = os.sep.join([read_path, fw_name])\n",
-        "fw_shape = Coordinates.from_json(fw_name)\n",
-        "=======\n",
         "fw_name = os.path.join(read_path, fw_name)\n",
-        "fw_shape = Loop.from_file(fw_name)\n",
-        ">>>>>>> 44bf57e9 (Remove ParameterFrame from radiation_transport module (#1472))"
+        "fw_shape = Coordinates.from_json(fw_name)"
```
Comment (hsaunders1904) [inline examples/radiation_transport/heat_flux_calculation_DN.ipynb:55]: Whoops! Thanks for fixing this

File: tests/codes/process/test_setup.py [hunk 11]
```diff
@@ -38,19 +38,18 @@ class TestSetup:
     def setup_method(self):
         self.default_pf = ProcessSolverParams.from_json(PARAM_FILE)
 
-        self._writer_patch = mock.patch(f"{MODULE_REF}.InDat")
-        self.writer_cls_mock = self._writer_patch.start()
-
-    def teardown_method(self):
-        self._writer_patch.stop()
```
Comment (hsaunders1904) [inline tests/codes/process/test_setup.py:45]: I think we still need teardowns for the patches, or they're going to persist past the life of the class, and (potentially) mess with subsequent tests.

File: tests/codes/process/test_teardown.py [hunk 12]
```diff
@@ -69,7 +69,7 @@ def test_read_func_updates_bluemira_params_from_mfile(self, run_func, result):
         # Expected value comes from ./test_data/mfile_data.json
         assert teardown.params.tau_e.value == pytest.approx(4.3196)
         # auto unit conversion
-        assert teardown.params.P_el_net.value == pytest.approx(result)
+        assert teardown.params.P_el_net.value == pytest.approx(5e8)
```
Comment (hsaunders1904) [inline tests/codes/process/test_teardown.py:72]: Has this changed because of our units standardisation?

File: tests/codes/test_param.py [hunk 13]
```diff
@@ -55,3 +58,66 @@ def test_value_change(self):
...
+    def from_defaults(cls) -> MappedParameterFrame:
+        default_dict = asdict(cls._defaults)
+        default_dict["d"] = 0
+        default_dict["e"] = "0"
+        default_dict["f"] = False
+        return super().from_defaults(default_dict)
+
+
+class TestDefaultPM:
+    def test_mapped_default_pm_sets_unknown_values(self):
```
Comment (hsaunders1904) [inline tests/codes/test_param.py:107]: Would `unmapped` be more descriptive than `unknown` here? Similar in the test below

**Review by hsaunders1904 — CHANGES_REQUESTED** [review]
Lots of good stuff here, definitely a lot more to think about than I'd first thought!

File: bluemira/codes/params.py [hunk 14]
```diff
@@ -40,6 +40,29 @@ class MappedParameterFrame(ParameterFrame):
...
+    def from_defaults(cls, data: Dict) -> MappedParameterFrame:
+        """
+        Create ParameterFrame with default values for external codes.
+
+        External codes are likely to have variables that are not changed often
+        therefore in some cases sane defaults are needed
+
+        """
+        new_param_dict = {}
+        for map_name, param_map in cls._mappings.items():
```
[the remark this answers is not in the document]
Comment (je-cook) [reply bluemira/codes/params.py:59]: map_name is the bluemira name, I will change it

File: bluemira/codes/params.py [hunk 15]
```diff
@@ -40,6 +40,29 @@ class MappedParameterFrame(ParameterFrame):
...
+            new_param_dict[map_name] = {
+                "value": data[param_map.name],
+                "unit": param_map.unit,
+            }
```
[the remark this answers is not in the document]
Comment (je-cook) [reply bluemira/codes/params.py:63]: there probably wont be a description or a long name but I will add a source

File: bluemira/codes/plasmod/api/_solver.py [hunk 16]
```diff
@@ -94,7 +94,8 @@ def __init__(
         self._run: Run
         self._teardown: Teardown
 
-        self.params = make_parameter_frame(params, PlasmodSolverParams)
+        self.params = PlasmodSolverParams.from_defaults()
+        self.params.update_from_frame(params)
```
[the remark this answers is not in the document]
Comment (je-cook) [reply bluemira/codes/plasmod/api/_solver.py:98]: ahh yeah it wont...

File: bluemira/codes/process/_inputs.py (hunk 4 above)
Comment (je-cook) [reply bluemira/codes/process/_inputs.py:70]: For all of these the units are in the mappings or they live in a process dictionary. The process units are a whole other issue...they've been working on it for a while

File: bluemira/codes/process/_inputs.py (hunk 5 above)
Comment (je-cook) [reply bluemira/codes/process/_inputs.py:261]: When this is first called they are not INVariables (the start of the post_init) and then I use the field name to create the INVariable.

File: bluemira/codes/process/_inputs.py (hunk 4 above)
Comment (je-cook) [reply bluemira/codes/process/_inputs.py:70]: added a bit to the description

File: bluemira/codes/plasmod/api/_solver.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (je-cook) [reply bluemira/codes/plasmod/api/_solver.py]: modified and test added

File: bluemira/codes/process/_inputs.py [hunk 17]
```diff
@@ -36,6 +36,9 @@ class ProcessInputs:
     Notes
     -----
     All entries get wrapped in an INVariable class to enable easy InDat writing
```
Comment (hsaunders1904) [inline bluemira/codes/process/_inputs.py:38]:
```suggestion
    All entries get wrapped in an INVariable class to enable easy InDat writing.
    
```

File: bluemira/codes/process/_setup.py [hunk 18]
```diff
@@ -130,10 +130,11 @@ def _validate_models(self, writer):
...
 
 
-def _make_writer(template_in_dat: Union[str, ProcessInputs]) -> InDat:
-    if isinstance(template_in_dat, ProcessInputs):
+def _make_writer(template_in_dat: Union[str, Dict[str, _INVariable]]) -> InDat:
+    if isinstance(template_in_dat, Dict):
         indat = InDat(filename=None)
-        indat.data = template_in_dat.to_dict()
+        indat.data = template_in_dat
+        print(indat.data)
```
Comment (hsaunders1904) [inline bluemira/codes/process/_setup.py:137]:
```suggestion
```

File: tests/codes/plasmod/test_solver.py [hunk 19]
```diff
@@ -60,12 +60,17 @@ def test_inputs_updated_from_problem_settings_on_init(self):
...
         new_inputs = {
             "v_loop": -1.5e-3,
             "q_heat": 1.5,
             "nx": 25,
         }
+        if param_type == "dict":
+            param = {k: data["value"] for k, data in self.default_pf.to_dict().items()}
+        else:
+            param = self.default_pf
         setup = Setup(self.default_pf, {}, self.input_file)
```
Comment (hsaunders1904) [inline tests/codes/plasmod/test_solver.py:74]: Is this new `param` variable used?

Did you mean this?
```suggestion
        if param_type == "dict":
            params = {k: data["value"] for k, data in self.default_pf.to_dict().items()}
        else:
            params = self.default_pf
        setup = Setup(params, {}, self.input_file)
```

**Review by hsaunders1904 — APPROVED** [review]
This looks good to me now ✔️

Just a couple of minor suggestions.

Good job getting this working, I appreciate the issue was more complex than we first thought. Fingers crossed that this becomes easier once PROCESS's Python API is properly available!

File: tests/codes/plasmod/test_solver.py (hunk 19 above)
Comment (je-cook) [reply tests/codes/plasmod/test_solver.py]: Yes but its in the wrong place either way, the last commit fixes it

**Pull request merged**